[Asher Lim] iP

doimoiboi opened the pull request:
# Duke:DoiMoiBot

## Latest and Greatest
Just like the original, Duke:DoiMoiBot frees your mind of having to remember!
- Text-based
- Easy to learn
- Fast to use

## Installation
__Get your FREE download now!__
1. Download it from [here](https://github.com/doimoiboi/ip).
2. Run the program.
3. Add your tasks.
4. Manage your tasks.
5. ???
6. Profit :moneybag: :moneybag: :moneybag:

## Quick Start
Some basic commands are:

```
todo 
mark
delete
```

## Features
Current Features:
- [x] Managing by Tasks
- [ ] Managing by Deadline (coming soon)
- [ ] Reminders (coming soon)

## Satisfied Customers
> "This. This is the reason for my success" - Yi Long Ma
> "I use Duke:DoiMoiBot 5 times a day 10 times a day." - Heffery Bayzoes
> "I hope Duke IPOs so that I can hold it long" - Darren Muffet

## For Developer Eyes Only...
Developers can use it to practice their Java too! The `main` method is as follows:

```ruby
    public static void main(String[] args) {
        Duke duke = new Duke("./data/duke.txt");
        duke.run();
    }
```

doimoiboi committed:
Added Level 1 features

doimoiboi committed:
Completed Level 2

doimoiboi committed:
Completed level 3

doimoiboi committed:
completed Level 4

doimoiboi committed:
done automated text UI testing

doimoiboi committed:
Modified code to only accept list commands, exceptions were already created previously

doimoiboi committed:
Added delete functionality

doimoiboi committed:
Duke now able to save data to hard disk and modify data

doimoiboi committed:
Add ability for duke to recognise dates

doimoiboi committed:
Merge branch 'branch-Level-7'

doimoiboi committed:
Merge branch 'branch-Level-8'

doimoiboi committed:
Refactored code to include better OOP practices.

doimoiboi committed:
Organised git classes into packages

doimoiboi committed:
Add Gradle support

doimoiboi committed:
Add JUnit test classes

doimoiboi committed:
Added search functionality

doimoiboi committed:
Add JavaDoc header comments

doimoiboi committed:
Edit code to comply with coding standards

doimoiboi committed:
Merge branch 'branch-A-CodingStandard'

doimoiboi committed:
Merge branch 'branch-A-JavaDoc'

doimoiboi committed:
Merge branch 'branch-Level-9'

lleyi0606 commented on src/main/java/duke/task/Deadlines.java:
```diff
@@ -0,0 +1,99 @@
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+public class Deadlines extends Task {
```
Might be better to name this class as `Deadline` instead as it is meant to represent a single deadline task

lleyi0606 commented on src/main/java/duke/task/Events.java:
```diff
@@ -0,0 +1,104 @@
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+public class Events extends Task {
```
Same as `Deadlines` class. Might be better to name as `Event` 😄

lleyi0606 commented on src/main/java/duke/task/ToDos.java:
```diff
@@ -0,0 +1,51 @@
+package duke.task;
+
+public class ToDos extends Task {
```
Same as `Deadlines` class. Might be better to name as `ToDo` 😃

lleyi0606 commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,18 @@
+public class ParserTest {
+
+    @Test
+    public void dataToInfo_ToDosTask_success() {
```
Very up-to-standard naming for tests! 💯

lleyi0606 reviewed (commented):
Very clean and organised code! Can't find much to comment on 😆

nopehax commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,149 @@
+     */
+    public static Command parse(String input) throws UnknownCommandException {
+        try {
+            if (input.equals("bye")) {
+                return new ExitCommand();
+            }
+
+            if (input.equals("list")) {
+                return new ListCommand();
+            }
+            //logic to mark tasks with error handling
+            if (input.indexOf("mark") == 0) {
+                String substring = input.substring(5);
+                //Since first task is of index 0 in ArrayList
+                int taskIndex = Integer.parseInt(substring) - 1;
+                return new MarkCommand(taskIndex, true);
+            }
+            //logic to unmark tasks with error handling
+            if (input.indexOf("unmark") == 0) {
+                String substring = input.substring(7);
+                //Since first task is of index 0 in ArrayList
+                int taskIndex = Integer.parseInt(substring) - 1;
+                return new MarkCommand(taskIndex, false);
+            }
+            //logic to create duke.task.ToDos
+            if (input.indexOf("todo") == 0) {
+                String description = input.substring(5);
+                ToDos todo = new ToDos(description, false);
+                return new AddCommand(todo);
+            }
+            //logic to create duke.task.Deadlines
+            if (input.indexOf("deadline") == 0) {
+                //remove initial command
+                String sub = input.substring(9);
+                int timeIndex = sub.lastIndexOf("/by");
+                if (timeIndex == -1) {
+                    throw new MissingDeadlineException();
+                }
```
Is there a reason why you use if else statements instead of switch statements? Maybe you can use enums for the different commands

nopehax commented on src/main/java/duke/task/Deadlines.java:
```diff
@@ -0,0 +1,99 @@
+//                throw new duke.dukeexception.MissingDeadlineException();
+//            }
+//            String description = sub.substring(0, timeIndex - 1);
+//            this.description = description;
+//            String deadlineString = sub.substring(timeIndex + 4);
+//            LocalDateTime deadline = LocalDateTime.parse(deadlineString, DATE_TIME_INPUT_FORMAT);
+//            this.deadline = deadline;
+//        } catch (StringIndexOutOfBoundsException e) {
+//            throw new duke.dukeexception.MissingDescriptionException();
+//        }
+//    }
+//
+//    public duke.task.Deadlines(String input, boolean isDone)
+//            throws duke.dukeexception.MissingDescriptionException, duke.dukeexception.MissingDeadlineException, DateTimeParseException {
+//        super(isDone);
+//        try {
+//            //remove initial command
+//            String sub = input.substring(9);
+//            int timeIndex = sub.lastIndexOf("/by");
+//            //get description part of input string
+//            if (timeIndex == -1) {
+//                throw new duke.dukeexception.MissingDeadlineException();
+//            }
+//            String description = sub.substring(0, timeIndex - 1);
+//            this.description = description;
+//            String deadlineString = sub.substring(timeIndex + 4);
+//            LocalDateTime deadline = LocalDateTime.parse(deadlineString, DATE_TIME_INPUT_FORMAT);
+//            this.deadline = deadline;
+//        } catch (StringIndexOutOfBoundsException e) {
+//            throw new duke.dukeexception.MissingDescriptionException();
+//        }
+//    }
+//
+//    public duke.task.Deadlines(String description, String deadlineString, boolean isDone) throws DateTimeParseException {
+//        super(isDone);
+//        this.description = description;
+//        LocalDateTime deadline = LocalDateTime.parse(deadlineString, DATE_TIME_INPUT_FORMAT);
+//        this.deadline = deadline;
+//    }
+}
```
Is there a reason why you didn't delete this chunk of code? It makes your code a bit more messy

nopehax reviewed (commented):
Overall good job! I like how you have JavaDocs for all, if not most, of your methods. Why not add JavaDocs for the different classes too

doimoiboi committed:
Halfway through implementing GUI

AkkFiros commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,149 @@
+                String search = input.substring(5);
+                return new FindCommand(search);
+            }
+        } catch (StringIndexOutOfBoundsException e) {
+            System.out.println("Please indicate which task to mark!");
+        } catch (NumberFormatException e) {
+            System.out.println("Please input a valid task index!");
+        } catch (IndexOutOfBoundsException e) {
+            System.out.println("duke.task.Task with that index does not exist!");
+        } catch (MissingTimingException e) {
+            System.out.println("Please specify the timeline!");
+        } catch (DateTimeParseException e) {
+            System.out.println("Please input a valid time in the format DD/MM/YYYY HHMM");
+        } catch (MissingDeadlineException e) {
+            System.out.println("Please specify the deadline!");
+        }
```
Could this block be abstracted into an exception class?

AkkFiros reviewed (commented):
Code is very clean and very easy to follow! The inclusion of JavaDocs for all (if not most) of the methods and classes really helps in the readability!

doimoiboi committed:
Add Duke GUI albeit code is messy.

doimoiboi committed:
Improve Code Quality.

- Remove unused code

- Edit javadoc comments

- Abstract code to parse input description for Deadlines and Events.

doimoiboi committed:
Add assertions

-Assertion in Duke.java to assert that task list is not null in Duke

doimoiboi committed:
Fix bugs.

-Add empty Duke constructor for launcher to work

doimoiboi committed:
Merge pull request #2 from doimoiboi/branch-A-CodeQuality

Improve code quality

doimoiboi committed:
Merge branch 'master' into branch-A-Assertions

doimoiboi committed:
Merge pull request #3 from doimoiboi/branch-A-Assertions

Add assertions

doimoiboi committed:
Add DoAfter task type

-Add DoAfter task type

doimoiboi committed:
Update user guide in README.md

doimoiboi committed:
Minor Code Quality changes

- Remove unnecessary code

- Correct lack of SLAP

- Update User Guide in README.me

doimoiboi committed:
Add screenshot for user guide

doimoiboi committed:
Link screenshot to user guide

Commit:
Fix issue where directory was not created and hence jar storage would fail

doimoiboi committed:
Fix user guide screenshot not showing and delete command bug.
